diff --git a/cmake/dependency_utils.cmake b/cmake/dependency_utils.cmake
index 8c54880..89fb04a 100644
--- a/cmake/dependency_utils.cmake
+++ b/cmake/dependency_utils.cmake
@@ -23,8 +23,8 @@ endmacro()
 
 macro(_use_system_libraries)
 	# Use system libraries
-	_import_aer_system_dependency(nlohmann_json 3.1.1)
-	_import_aer_system_dependency(spdlog 1.5.0)
+	#_import_aer_system_dependency(nlohmann_json 3.1.1)
+	#_import_aer_system_dependency(spdlog 1.5.0)
 
 	if(SKBUILD)
 		_import_aer_system_dependency(muparserx 4.0.8)
